src/eip712.rs

```diff
@@ -0,0 +1,571 @@
+                type_: Type::BytesN(32)
+            }
+        ])
+    };
+}
+
+impl Struct {
+    /// [`encodeType`](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-712.md#definition-of-encodetype)
+    pub fn encode(&self, name: &str, more_types: &Types) -> Result<Vec<u8>, TypedDataHashError> {
+        let mut string = String::new();
+        self.encode_single(name, &mut string);
+        let mut referenced_types = HashMap::new();
+        self.gather_referenced_types(more_types, &mut referenced_types)?;
+        let mut types: Vec<(&String, &Struct)> = referenced_types.into_iter().collect();
+        types.sort_by(|(name1, _), (name2, _)| name1.cmp(name2));
+        for (name, type_) in types {
+            type_.encode_single(name, &mut string);
+        }
+        Ok(string.into_bytes())
+    }
+
+    fn encode_single(&self, name: &str, string: &mut String) {
+        string.push_str(&name);
+        string.push('(');
+        let mut first = true;
+        for member in &self.0 {
+            if first {
+                first = false;
+            } else {
+                string.push(',');
+            }
+            string.push_str(&String::from(member.type_.clone()));
+            string.push(' ');
+            string.push_str(&member.name);
+        }
+        string.push(')');
+    }
+
+    fn gather_referenced_types<'a>(
+        &'a self,
+        types: &'a Types,
+        memo: &mut HashMap<&'a String, &'a Struct>,
+    ) -> Result<(), TypedDataHashError> {
+        for member in &self.0 {
+            if let Type::Reference(ref reference_name) = member.type_ {
+                use std::collections::hash_map::Entry;
+                let entry = memo.entry(reference_name);
+                if let Entry::Vacant(o) = entry {
+                    let referenced_struct = types.types.get(reference_name).ok_or(
+                        TypedDataHashError::MissingReferencedType(reference_name.to_string()),
+                    )?;
+                    o.insert(referenced_struct);
+                    referenced_struct.gather_referenced_types(types, memo)?;
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+impl Data {
+    /// [`hashStruct`](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-712.md#definition-of-hashstruct)
+    pub fn hash(
+        &self,
+        type_: &Struct,
+        name: &str,
+        more_types: &Types,
+    ) -> Result<[u8; 32], TypedDataHashError> {
+        let encoded_data = self.encode(type_, more_types)?.to_vec();
+        let encoded_type = type_.encode(name, more_types)?;
+        let type_hash = keccak(encoded_type).to_fixed_bytes().to_vec();
+        Ok(keccak(vec![type_hash, encoded_data].concat()).to_fixed_bytes())
+    }
+
+    /// [`encodeData`](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-712.md#definition-of-encodedata)
+    pub fn encode(&self, _type: &Struct, _types: &Types) -> Result<Vec<u8>, TypedDataHashError> {
```

Comment:
Just to confirm, will we need to complete this section for verification?

Reply:
Yes, unless using another EIP-712 implementation.